packages/devtools_app/lib/src/code_size/code_size_controller.dart

```diff
@@ -82,113 +71,119 @@ class CodeSizeController {
+  /// represents a change in size for the given node.
+  ///
+  /// The tree can be filtered with different [DiffTreeType] values:
+  /// * [DiffTreeType.increaseOnly]: returns a tree with nodes with positive [byteSize].
+  /// * [DiffTreeType.decreaseOnly]: returns a tree with nodes with negative [byteSize].
+  /// * [DiffTreeType.combined]: returns a tree with all nodes.
   TreemapNode generateDiffTree(
     Map<String, dynamic> treeJson, {
     DiffTreeType diffTreeType = DiffTreeType.combined,
   }) {
-    var treemapNodeName = treeJson['n'];
-    if (treemapNodeName == '') treemapNodeName = 'Unnamed';
-    final rawChildren = treeJson['children'];
-    final treemapNodeChildren = <TreemapNode>[];
-
-    int treemapNodeSize = 0;
-    if (rawChildren != null) {
-      // If not a leaf node, build all children then take the sum of the
-      // children's sizes as its own size.
-      for (dynamic child in rawChildren) {
-        final childTreemapNode = generateDiffTree(
-          child,
-          diffTreeType: diffTreeType,
-        );
-        if (childTreemapNode == null) {
-          continue;
-        }
-        treemapNodeChildren.add(childTreemapNode);
-        treemapNodeSize += childTreemapNode.byteSize;
-      }
+    final isLeafNode = treeJson['children'] == null;
+    if (!isLeafNode) {
+      return buildNodeWithChildren(
+        treeJson,
+        showDiff: true,
+        diffTreeType: diffTreeType,
+      );
     } else {
-      // If a leaf node, just take its own size.
-      treemapNodeSize = treeJson['value'];
-
-      // Only add nodes with a size.
-      if (treemapNodeSize == null || treemapNodeSize == 0) {
+      // TODO(peterdjlee): Investigate why there are leaf nodes with size of null.
+      final byteSize = treeJson['value'];
+      if (byteSize == null) {
         return null;
       }
 
       // Only add nodes that match the diff tree type.
       switch (diffTreeType) {
         case DiffTreeType.increaseOnly:
-          if (treemapNodeSize < 0) {
+          if (byteSize < 0) {
             return null;
           }
           break;
         case DiffTreeType.decreaseOnly:
-          if (treemapNodeSize > 0) {
+          if (byteSize > 0) {
             return null;
           }
           break;
         case DiffTreeType.combined:
           break;
       }
+
+      return buildNode(treeJson, byteSize, showDiff: true);
     }
+  }
 
+  /// Builds a node by recursively building all of its children first
+  /// in order to calculate the sum of its children's sizes.
+  TreemapNode buildNodeWithChildren(
+    Map<String, dynamic> treeJson, {
+    bool showDiff = false,
+    DiffTreeType diffTreeType = DiffTreeType.combined,
+  }) {
+    final rawChildren = treeJson['children'];
+    final treemapNodeChildren = <TreemapNode>[];
+    int totalByteSize = 0;
+
+    // Given a child, build its subtree.
+    for (dynamic child in rawChildren) {
```

Comment:
can we use something other than dynamic? is each child a map?

Reply:
Fixed